File: ActionUI/Views/Color.swift
// Sources/Views/Color.swift
/*
 Sample JSON for Color:
 {
   "type": "Color",
   "id": 1,              // Optional: Non-zero positive integer for runtime programmatic interaction
   "properties": {
     "color": "blue"   // Required: a color string - a named color ("red", "primary"/"secondary"),
                        // a hex string ("#RRGGBB" / "#RRGGBBAA"), or "<color>.opacity(<fraction>)".
   }
   // SwiftUI's `Color` is itself a View: a greedy block that fills the proposed space - the
   // canonical way to use a solid color as a view (a colored divider, a tinted background block),
   // simpler than a Rectangle().fill(). Give it a `frame` to size it (e.g. height 2 for a divider).
   // Baseline View properties (frame, padding, opacity, cornerRadius, etc.) are inherited via
   // ActionUIRegistry.shared.applyViewModifiers.
 }
 // Observable state:
 //   value (SwiftUI.Color)  A runtime color override - nil means "use the `color` property". Settable via
 //                          setElementValue (a Color) or setElementValueFromString (a color string - "red",
 //                          "#FF0000", "blue.opacity(0.3)"). ActionUIView re-builds on the change, so it re-renders.
 NOTE: this `Color` is ActionUI's element type (registered as "Color"). It is a module-level type,
 so it shadows `SwiftUI.Color` for unqualified uses within ActionUI - the few SwiftUI color sites
 (ColorHelper, ColorPicker, ActionUIModel) are written as `SwiftUI.Color` to disambiguate.
*/

import SwiftUI

struct Color: ActionUIViewConstruction {
    // Optional SwiftUI.Color value: nil = use the `color` property, a Color = a runtime override (set via
    // setElementValue / setElementValueFromString). Mirrors Image's static-property-with-runtime-override shape.
    static var valueType: Any.Type = SwiftUI.Color?.self
    static var initialStates: (ViewModel) -> [String: Any] = { model in model.states }
    static var parseStringValue: ((String, String?, any ActionUILogger) -> Any?)? = nil
    static var serializeValueToString: ((Any, String?, any ActionUILogger) -> String?)? = nil
    static var insertableContainers: [String: ContainerShape]? = nil

    static var validateProperties: ([String: Any], any ActionUILogger) -> [String: Any] = { properties, logger in
        var validated = properties
        if let color = validated["color"], !(color is String) {
            logger.log("Color color must be a String; ignoring", .warning)
            validated["color"] = nil
        }
        return validated
    }

    static var initialValue: (ViewModel) -> Any? = { model in
        // Non-nil only when a color was set at runtime; nil means "use the `color` property" (buildView resolves it).
        return model.value as? SwiftUI.Color
    }

    static var buildView: (any ActionUIElementBase, ViewModel, String, [String: Any], any ActionUILogger) -> any SwiftUI.View = { _, model, _, properties, logger in
        // Runtime override takes precedence (ActionUIView re-builds this when model.value changes).
        if let color = model.value as? SwiftUI.Color {
            return color
        }
        let colorStr = properties["color"] as? String
        if let colorStr, !colorStr.isEmpty {
            if let resolved = ColorHelper.resolveColor(colorStr) {
                return resolved
            }
            logger.log("Unknown color \"\(colorStr)\" for Color view; rendering clear", .warning)
        } else {
            logger.log("Color view requires a 'color' string; rendering clear", .warning)
        }
        return SwiftUI.Color.clear
    }

    static var applyModifiers: (any SwiftUI.View, any ActionUIElementBase, String, [String: Any], any ActionUILogger) -> any SwiftUI.View = { view, _, _, _, _ in
        return view
    }
}

File: ActionUIWeb/src/Views/Color.js
// Color.js — Color element (SwiftUI's `Color` used as a View).
// Web analog of ActionUI/Views/Color.swift and ActionUIAndroid Views/Color.kt.
//
// SwiftUI's `Color` is itself a View: a greedy block that fills the proposed space - the
// canonical way to use a solid color as a view (a colored divider, a tinted background block),
// simpler than a Rectangle().fill(). It reuses the shape box (Helpers/ShapeStyleHelper.js):
// greedy without a frame (`.aui-shape`), pinned with one. Give it a `frame` to size it
// (e.g. height 2 for a divider). Baseline View properties are inherited via ModifierResolver.
//
// Property: color (String) - a named color ("red"), an Apple semantic style
// ("primary"/"secondary"/"fill.tertiary"/"tint"/...), a hex string ("#RRGGBB" / "#RRGGBBAA"),
// or "<color>.opacity(<fraction>)". Resolved through the shared resolveColor (semantic names
// map to adaptive --aui-* tokens; it warns on a genuinely unknown color); a missing/unresolvable
// color renders transparent.

import { register } from "../Common/ActionUIRegistry.js";
import { resolveColor } from "../Common/ModifierResolver.js";
import { createShapeBox } from "../Helpers/ShapeStyleHelper.js";

register("Color", {
    valueType: "none",

    validateProperties: (properties, logger) => {
        const validated = { ...properties };
        if (validated.color !== undefined && typeof validated.color !== "string") {
            logger.log("Color color must be a String; ignoring", "warning");
            delete validated.color;
        }
        return validated;
    },

    buildView: (element, properties, ctx) => {
        const node = createShapeBox(properties, "aui-color");
        const colorStr = properties.color;
        if (typeof colorStr === "string" && colorStr.length > 0) {
            const css = resolveColor(colorStr, ctx.logger); // warns on an unknown color
            if (css) node.style.background = css;
        } else {
            ctx.logger.log("Color view requires a 'color' string; rendering transparent", "warning");
        }
        return node;
    },
});
